### DISCUSSION TITLE
Version of libgphoto installed on system

### DISCUSSION THREAD

**Discussion by mgineer85** [jim-easterbrook/python-gphoto2#172] [Q&A] [answered]
Hey, it seems `gp.gp_library_version(gp.GP_VERSION_VERBOSE)}` always returns the version of the installed library in the system.

If it is installed with --prefer-binaries the more recent included version is used, but the function still returns the system version information.

Is there any way to check if the precompiled version coming along with the latest python-gphoto2 is used?

**Comment by jim-easterbrook** [comment]
I didn't know about `--prefer-binaries`, it sounds as if it leaves scope for doing a source installation. I use `--only-binary :all:` to ensure a binary installation.

First find out where `python-gphoto2` is installed:
```
$ pip show gphoto2
Name: gphoto2
Version: 2.5.1
Summary: Python interface to libgphoto2
Home-page: https://github.com/jim-easterbrook/python-gphoto2
Author: Jim Easterbrook
Author-email: jim@jim-easterbrook.me.uk
License: GNU LGPL
Location: /home/jim/.local/lib/python3.6/site-packages
Requires: 
Required-by: 
```

Then check you have a `libgphoto2` directory installed with `python-gphoto2`:
```
$ ls -l /home/jim/.local/lib/python3.6/site-packages/gphoto2/libgphoto2/
total 636
drwxr-xr-x  2 jim users   4096 Nov  7  2023 camlibs
drwxr-xr-x  2 jim users   4096 Nov  7  2023 iolibs
-rwxr-xr-x  1 jim users 138872 Nov  7  2023 libgphoto2_port.so.12
-rwxr-xr-x  1 jim users 495392 Nov  7  2023 libgphoto2.so.6
drwxr-xr-x 24 jim users   4096 Nov  7  2023 locale
drwxr-xr-x  2 jim users   4096 Nov  7  2023 vusb
```

If this is all as it should be, then use `ldd` to ensure `python-gphoto2` is using the correct libraries:
```
$ ldd /home/jim/.local/lib/python3.6/site-packages/gphoto2/_version.cpython-36m-x86_64-linux-gnu.so
        linux-vdso.so.1 (0x00007ffc6a88c000)
        libgphoto2.so.6 => /home/jim/.local/lib/python3.6/site-packages/gphoto2/libgphoto2/libgphoto2.so.6 (0x00007f7ed1102000)
        libm.so.6 => /lib64/libm.so.6 (0x00007f7ed0f88000)
        libgphoto2_port.so.12 => /home/jim/.local/lib/python3.6/site-packages/gphoto2/libgphoto2/libgphoto2_port.so.12 (0x00007f7ed0f7b000)
        libpython3.6m.so.1.0 => /usr/lib64/libpython3.6m.so.1.0 (0x00007f7ed0c9b000)
        libpthread.so.0 => /lib64/libpthread.so.0 (0x00007f7ed0c75000)
        libc.so.6 => /lib64/libc.so.6 (0x00007f7ed0a7e000)
        libltdl.so.7 => /usr/lib64/libltdl.so.7 (0x00007f7ed0a73000)
        libdl.so.2 => /lib64/libdl.so.2 (0x00007f7ed0a6e000)
        libexif.so.12 => /usr/lib64/libexif.so.12 (0x00007f7ed0800000)
        /lib64/ld-linux-x86-64.so.2 (0x00007f7ed1192000)
        libutil.so.1 => /lib64/libutil.so.1 (0x00007f7ed0a68000)
```

**Reply by mgineer85** [reply]
I see, I will check later.
I would actually like to avoid prefer binaries (using for opencv and gphoto) though it seems to work fine.

Converting things to venvs nowadays doesn’t simplify things: https://photobooth-app.org/setup/installation/#install-photobooth-app

**Reply by mgineer85** [reply]
1) 
After `pip install --force  --prefer-binary gphoto2`

result of gphoto2_version.py from examples is
```
python: 3.9.2 (default, Feb 28 2021, 17:03:44)
[GCC 10.2.1 20210110]
libgphoto2: ['2.5.27', 'standard camlib set (ax203 canon digigr8 dimagev directory docupen jl2005a jl2005c kodak_dc240 lumix mars pentax ptp2 ricoh_g3 sierra sonix sq905 st2205 topfield tp6801)', 'gcc (C compiler used)', 'ltdl (for portable loading of camlibs)', 'EXIF (for special handling of EXIF files)']
libgphoto2_port: ['0.12.0', 'iolibs: disk ptpip serial usb1 usbdiskdirect usbscsi', 'gcc (C compiler used)', 'ltdl (for portable loading of iolibs)', 'EXIF (for vusb)', 'USB (libusb1, for USB cameras)', 'serial (for serial cameras)', 'no resmgr (serial port access and locking)', 'no ttylock (serial port locking)', 'no lockdev (serial port locking)']
python-gphoto2: 2.5.0
```

result of `pip show gphoto2` is
```
linaro@photobooth-blue:~$ pip3 show gphoto2
Name: gphoto2
Version: 2.5.0
Summary: Python interface to libgphoto2
Home-page: None
Author: None
Author-email: Jim Easterbrook <jim@jim-easterbrook.me.uk>
License: GNU LGPL
Location: /home/linaro/.local/lib/python3.9/site-packages
Requires:
Required-by: photobooth-app
```

indeed, the libraries were not downloaded correctly
```
linaro@photobooth-blue:~$ ls -l /home/linaro/.local/lib/python3.9/site-packages/gphoto2/libgphoto2
ls: Zugriff auf '/home/linaro/.local/lib/python3.9/site-packages/gphoto2/libgphoto2' nicht möglich: Datei oder Verzeichnis nicht gefunden
```


and system libraries are used, which seems to be 2.5.27 on my system.
```
linaro@photobooth-blue:~$ ldd /home/linaro/.local/lib/python3.9/site-packages/gphoto2/_version.cpython-39-aarch64-linux-gnu.so
        linux-vdso.so.1 (0x0000007f8a6c9000)
        libgphoto2.so.6 => /lib/aarch64-linux-gnu/libgphoto2.so.6 (0x0000007f8a5d1000)
        libgphoto2_port.so.12 => /lib/aarch64-linux-gnu/libgphoto2_port.so.12 (0x0000007f8a5b5000)
        libpthread.so.0 => /lib/aarch64-linux-gnu/libpthread.so.0 (0x0000007f8a584000)
        libc.so.6 => /lib/aarch64-linux-gnu/libc.so.6 (0x0000007f8a410000)
        /lib/ld-linux-aarch64.so.1 (0x0000007f8a699000)
        libltdl.so.7 => /lib/aarch64-linux-gnu/libltdl.so.7 (0x0000007f8a3f6000)
        libexif.so.12 => /lib/aarch64-linux-gnu/libexif.so.12 (0x0000007f8a39d000)
        libm.so.6 => /lib/aarch64-linux-gnu/libm.so.6 (0x0000007f8a2f2000)
        libdl.so.2 => /lib/aarch64-linux-gnu/libdl.so.2 (0x0000007f8a2de000)
```

2) 
After `pip install --force gphoto2`

results all the same as for 1).


3)
After `pip install --force --only-binary :all: gphoto2`
```
linaro@photobooth-blue:~$ pip install gphoto2 --only-binary :all:
Defaulting to user installation because normal site-packages is not writeable
ERROR: Could not find a version that satisfies the requirement gphoto2 (from versions: none)
ERROR: No matching distribution found for gphoto2
```

So it seems it was never supported to use the prebuilt wheel using the precompiled binaries and the prefer-binaries covered it.
Strange I got camera connection issues earlier and not today any more with the installation made with --prefer-binaries


```
linaro@photobooth-blue:~$ uname -a
Linux photobooth-blue 5.10.160 #1 SMP Tue Oct 24 03:20:27 UTC 2023 aarch64 GNU/Linux
linaro@photobooth-blue:~$ lsb_release -d
Description:    Debian GNU/Linux 11 (bullseye)

```

Now I tried installing on a raspberry pi with Pi OS Bullseye:

```
pi@photobooth-yellow:~ $ pip3 install gphoto2 --user --only-binary :all:
Looking in indexes: https://pypi.org/simple, https://www.piwheels.org/simple
ERROR: Could not find a version that satisfies the requirement gphoto2
ERROR: No matching distribution found for gphoto2
```
All prerequisites mentioned here are installed: https://github.com/jim-easterbrook/python-gphoto2/blob/main/INSTALL.rst#raspberry-pi

Is there something I do wrong at this point?

**Reply by jim-easterbrook** [reply]
I've just spotted that you're using Python built for `aarch64-linux`. I don't currently provide a built distribution for that architecture, so binary installation isn't possible. (You can see the 14 supported architectures here: https://pypi.org/project/gphoto2/#files) I'll have to wait until GitHub provide a suitable "runner" before I can provide a built distribution.

As for the Raspberry Pi, is that also `aarch64`? See the note on https://www.piwheels.org/project/gphoto2/ "piwheels does not currently support aarch64".

**Reply by mgineer85** [reply]
Hi,

that explains. I thought with the PI I am safe and can use the latest binary versions. Did not expect its not supported ^^

Check the output for two systems, Pi4 and Pi5. I use the 64bit version of their system exclusively.

Pi4:
```
pi@photobooth-yellow:~ $ cat /proc/device-tree/model
Raspberry Pi 4 Model B Rev 1.4

pi@photobooth-yellow:~ $ lsb_release -a
No LSB modules are available.
Distributor ID: Debian
Description:    Debian GNU/Linux 11 (bullseye)
Release:        11

pi@photobooth-yellow:~ $ uname -a
Linux photobooth-yellow 6.1.21-v8+ #1642 SMP PREEMPT Mon Apr  3 17:24:16 BST 2023 aarch64 GNU/Linux
```

Pi5:
```
pi@photobooth-white:~ $ cat /proc/device-tree/model
Raspberry Pi 5 Model B Rev 1.0

pi@photobooth-white:~ $ lsb_release -a
No LSB modules are available.
Distributor ID: Debian
Description:    Debian GNU/Linux 12 (bookworm)
Release:        12
Codename:       bookworm

pi@photobooth-white:~ $ uname -a
Linux photobooth-white 6.6.28+rpt-rpi-2712 #1 SMP PREEMPT Debian 1:6.6.28-1+rpt1 (2024-04-22) aarch64 GNU/Linux
```

**Reply by mgineer85** [reply]
https://runs-on.com/ has arm64 runner to integrate with actions and is free for open source projects. Could be useful until actions supports arm64 natively by end of this year.

**Reply by jim-easterbrook** [reply]
Thanks for that, I'll look into it. I've used a non-GitHub runner for ARM64 MacOS builds on another project - it took some effort to get everything to build.

**Reply by mgineer85** [reply]
For the photobooth-app I use a Raspberry Pi as selfhosted runner. Works quite well and was not too much work to integrate in the actions environment. But sure, it's another device that needs maintenance and consumes power.

**Reply by jim-easterbrook** [reply]
The "run-on.com" runners run on Amazon's App Runner service, which as far as I can tell is not part of the free tier offering. (I have no arm64 machine to tun a runner on.)

You can install `python-gphoto2` to use a locally compiled (but not system installed) `libgphoto2`. See https://github.com/jim-easterbrook/python-gphoto2/blob/main/INSTALL.rst#local-libgphoto2 for details.

**Reply by mgineer85** [reply]
Hey @jim-easterbrook , GitHub has since some time arm runner in public beta available: https://docs.github.com/en/actions/using-github-hosted-runners/using-github-hosted-runners/about-github-hosted-runners#standard-github-hosted-runners-for-public-repositories

maybe building the arm binaries could be revisited with these available?

**Reply by jim-easterbrook** [reply]
Yes, I use it and have included MacOS arm64 binaries since python-gphoto2 version 2.5.1, released in October 2024.

**Reply by mgineer85** [reply]
So could Linux arm64 now also be added?

**Reply by jim-easterbrook** [reply]
Yes, it probably could. I didn't know there was much demand for it. I'm currently building and uploading to PyPI 22 different wheels. On a modern computer installing from source with `pip` is pretty quick, the only problem is getting inexperienced users to install the `-dev` package with their OS package manager.

**Reply by jim-easterbrook** [reply, marked as answer]
I've uploaded some arm64 wheels to PyPI.
